GPU/GLES/ShaderManager.cpp

```diff
@@ -944,3 +945,117 @@ std::string ShaderManager::DebugGetShaderString(std::string id, DebugShaderType
 		return "N/A";
 	}
 }
+
+// Shader pseudo-cache.
+//
+// We simply store the IDs of the shaders used during gameplay. On next startup of
+// the same game, we simply compile all the shaders from the start, so we don't have to
+// compile them on the fly later. Ideally we would store the actual compiled shaders
+// rather than just their IDs, but OpenGL does not support this, except for a few obscure
+// vendor-specific extensions.
+//
+// If things like GPU supported features have changed since the last time, we discard the cache
+// as sometimes these features might have an effect on the ID bits.
+
+#define CACHE_HEADER_MAGIC 0x83277592
+#define CACHE_VERSION 1
+struct CacheHeader {
+	uint32_t magic;
+	uint32_t version;
+	uint32_t featureFlags;
+	uint32_t reserved;
+	int numVertexShaders;
+	int numFragmentShaders;
+	int numLinkedPrograms;
+};
+
+void ShaderManager::LoadAndPrecompile(const std::string &filename) {
+	FILE *f = File::OpenCFile(filename, "rb");
+	if (!f) {
+		return;
+	}
+	CacheHeader header;
+	if (fread(&header, 1, sizeof(header), f) != sizeof(header)) {
+		fclose(f);
+		return;
+	}
+	if (header.magic != CACHE_HEADER_MAGIC || header.version != CACHE_VERSION || header.featureFlags != gstate_c.featureFlags) {
+		fclose(f);
+		return;
+	}
+	for (int i = 0; i < header.numVertexShaders; i++) {
+		ShaderID id;
+		fread(&id, 1, sizeof(id), f);
+		Shader *vs = CompileVertexShader(id);
+		if (vs->Failed()) {
+			// Give up on using the cache, just bail. We can't safely create the fallback shaders here
+			// without trying to deduce the vertType from the VSID.
+			ERROR_LOG(G3D, "Failed to compile a vertex shader loading from cache. Skipping rest of shader cache.");
+			delete vs;
+			fclose(f);
+			return;
+		}
+		vsCache_[id] = vs;
```

Comment:
How long does this take with lots of shaders?  Surely it can be done in a later pull, but just wondering if my idea about doing one per frame has merit or is a waste of time...

-[Unknown]